components-api/src/test/scala/pl/touk/nussknacker/engine/api/typed/AssignabilityDeterminerSpec.scala

```diff
@@ -0,0 +1,67 @@
+package pl.touk.nussknacker.engine.api.typed
+
+import cats.data.NonEmptyList
+import cats.data.Validated.{Invalid, Valid}
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.prop.TableDrivenPropertyChecks.forAll
+import org.scalatest.prop.Tables.Table
+import pl.touk.nussknacker.engine.api.typed._
+
+class AssignabilityDeterminerSpec extends AnyFunSuite with Matchers {
+
+  // Test data table: (sourceType, targetType, expectedStrict, expectedLoose)
+  val strictConversionCases = Table(
+    ("sourceType", "targetType", "expectedStrict", "expectedLoose"),
+    (typing.Typed[Int], typing.Typed[Int], Valid(()), Valid(())),                           // Same primitive type
+    (typing.Typed[Int], typing.Typed[Double], Valid(()), Valid(())),                        // Primitive widening
+    (typing.Typed[List[Int]], typing.Typed[List[Int]], Valid(()), Valid(())),               // Same generic type
+    (typing.Typed[List[Int]], typing.Typed[List[Any]], Valid(()), Valid(())),               // Generic type variance
+    (typing.Typed[Map[String, Int]], typing.Typed[Map[String, Int]], Valid(()), Valid(())), // Same map type
+    (typing.Typed[Map[String, Int]], typing.Typed[Map[Any, Any]], Valid(()), Valid(()))
+  ) // Different records
+
+  test("isAssignableStrict should pass for strict cases") {
+    forAll(strictConversionCases) { (sourceType, targetType, expectedStrict, _) =>
+      val result = AssignabilityDeterminer.isAssignableStrict(sourceType, targetType)
+      result shouldBe expectedStrict
+    }
+  }
+
+  test("isAssignableLoose should pass for strict cases") {
+    forAll(strictConversionCases) { (sourceType, targetType, _, expectedLoose) =>
+      val result = AssignabilityDeterminer.isAssignableLoose(sourceType, targetType)
+      result shouldBe expectedLoose
+    }
+  }
+
+  val looseConversionCases = Table(
+    ("sourceType", "targetType", "expectedStrict", "expectedLoose"),
+    (typing.Typed[Long], typing.Typed[Int], Invalid(NonEmptyList.of("")), Valid(())),
+    (typing.Typed[Double], typing.Typed[Int], Invalid(NonEmptyList.of("")), Invalid(NonEmptyList.of(""))),
+    (typing.Typed[BigDecimal], typing.Typed[Int], Invalid(NonEmptyList.of("")), Invalid(NonEmptyList.of("")))
+  )
+
+  test("isAssignableStrict should fail for looser numerical cases") {
+    forAll(looseConversionCases) { (sourceType, targetType, expectedStrict, _) =>
+      val result = AssignabilityDeterminer.isAssignableStrict(sourceType, targetType)
+      result match {
```

Comment:
This way of pattern matching in tests is not that readable and is harder to understand imo. For widening conversions both `isAssignableStrict` and `isAssignableLoose` should pass so we can just check if `isAssignableX` returns valid. For narrowing conversions `isAssignableLoose` should return valid and `isAssignableStrict` false.

Reply:
I think the pattern matching is ok, otherwise we would have to write specific match messages into the table and that cluttered it a lot.
I change the test cases so that the table is simpler.